{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Annotate somatic SVs using GENCODE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import necessary libraries for the analysis\n",
    "import os\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "from pathlib import Path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000290825.1;gene_id=ENSG00000290825.1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000456328.2;Parent=ENSG00000290825.1;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:1;Parent=ENST0000045...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:2;Parent=ENST0000045...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:3;Parent=ENST0000045...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname  source     feature  start    end score strand frame  \\\n",
       "0    chr1  HAVANA        gene  11869  14409     .      +     .   \n",
       "1    chr1  HAVANA  transcript  11869  14409     .      +     .   \n",
       "2    chr1  HAVANA        exon  11869  12227     .      +     .   \n",
       "3    chr1  HAVANA        exon  12613  12721     .      +     .   \n",
       "4    chr1  HAVANA        exon  13221  14409     .      +     .   \n",
       "\n",
       "                                           attribute  \n",
       "0  ID=ENSG00000290825.1;gene_id=ENSG00000290825.1...  \n",
       "1  ID=ENST00000456328.2;Parent=ENSG00000290825.1;...  \n",
       "2  ID=exon:ENST00000456328.2:1;Parent=ENST0000045...  \n",
       "3  ID=exon:ENST00000456328.2:2;Parent=ENST0000045...  \n",
       "4  ID=exon:ENST00000456328.2:3;Parent=ENST0000045...  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gencode = pd.read_table(\"/Users/ryanyutian/Desktop/annotation_dataset/gencode.v43.annotation.gff3\", comment=\"#\",\n",
    "                        sep = \"\\t\", names = ['seqname', 'source', 'feature', 'start' , 'end', 'score', 'strand', 'frame', 'attribute'])\n",
    "gencode.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['chr1', 'chr10', 'chr11', 'chr12', 'chr13', 'chr14', 'chr15',\n",
       "       'chr16', 'chr17', 'chr18', 'chr19', 'chr2', 'chr20', 'chr21',\n",
       "       'chr22', 'chr3', 'chr4', 'chr5', 'chr6', 'chr7', 'chr8', 'chr9',\n",
       "       'chrM', 'chrX', 'chrY'], dtype=object)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(gencode['seqname'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['CDS', 'exon', 'five_prime_UTR', 'gene', 'start_codon',\n",
       "       'stop_codon', 'stop_codon_redefined_as_selenocysteine',\n",
       "       'three_prime_UTR', 'transcript'], dtype=object)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(gencode['feature'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ID=ENSG00000223972.6;gene_id=ENSG00000223972.6;gene_type=transcribed_unprocessed_pseudogene;gene_name=DDX11L1;level=2;hgnc_id=HGNC:37102;havana_gene=OTTHUMG00000000961.2'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gencode.iloc[5]['attribute']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def gene_info(x):\n",
    "    \n",
    "    if 'Parent' in x:\n",
    "        \n",
    "        parent = list(filter(lambda x: 'Parent' in x,  x.split(\";\")))[0].split(\"=\")[1]\n",
    "        \n",
    "    else:\n",
    "        \n",
    "        parent = 'N/A'\n",
    "        \n",
    "    g_id = list(filter(lambda x: 'gene_id' in x,  x.split(\";\")))[0].split(\"=\")[1]\n",
    "    g_name = list(filter(lambda x: 'gene_name' in x,  x.split(\";\")))[0].split(\"=\")[1]\n",
    "    g_type = list(filter(lambda x: 'gene_type' in x,  x.split(\";\")))[0].split(\"=\")[1]\n",
    "    g_level = int(list(filter(lambda x: 'level' in x,  x.split(\";\")))[0].split(\"=\")[1])\n",
    "    \n",
    "    return (parent, g_id, g_name, g_type, g_level)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>parent</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000290825.1;gene_id=ENSG00000290825.1...</td>\n",
       "      <td>N/A</td>\n",
       "      <td>ENSG00000290825.1</td>\n",
       "      <td>DDX11L2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000456328.2;Parent=ENSG00000290825.1;...</td>\n",
       "      <td>ENSG00000290825.1</td>\n",
       "      <td>ENSG00000290825.1</td>\n",
       "      <td>DDX11L2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:1;Parent=ENST0000045...</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>ENSG00000290825.1</td>\n",
       "      <td>DDX11L2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:2;Parent=ENST0000045...</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>ENSG00000290825.1</td>\n",
       "      <td>DDX11L2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:3;Parent=ENST0000045...</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>ENSG00000290825.1</td>\n",
       "      <td>DDX11L2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname  source     feature  start    end score strand frame  \\\n",
       "0    chr1  HAVANA        gene  11869  14409     .      +     .   \n",
       "1    chr1  HAVANA  transcript  11869  14409     .      +     .   \n",
       "2    chr1  HAVANA        exon  11869  12227     .      +     .   \n",
       "3    chr1  HAVANA        exon  12613  12721     .      +     .   \n",
       "4    chr1  HAVANA        exon  13221  14409     .      +     .   \n",
       "\n",
       "                                           attribute             parent  \\\n",
       "0  ID=ENSG00000290825.1;gene_id=ENSG00000290825.1...                N/A   \n",
       "1  ID=ENST00000456328.2;Parent=ENSG00000290825.1;...  ENSG00000290825.1   \n",
       "2  ID=exon:ENST00000456328.2:1;Parent=ENST0000045...  ENST00000456328.2   \n",
       "3  ID=exon:ENST00000456328.2:2;Parent=ENST0000045...  ENST00000456328.2   \n",
       "4  ID=exon:ENST00000456328.2:3;Parent=ENST0000045...  ENST00000456328.2   \n",
       "\n",
       "             gene_id gene_name gene_type  gene_level  \n",
       "0  ENSG00000290825.1   DDX11L2    lncRNA           2  \n",
       "1  ENSG00000290825.1   DDX11L2    lncRNA           2  \n",
       "2  ENSG00000290825.1   DDX11L2    lncRNA           2  \n",
       "3  ENSG00000290825.1   DDX11L2    lncRNA           2  \n",
       "4  ENSG00000290825.1   DDX11L2    lncRNA           2  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gencode['parent'], gencode['gene_id'], gencode['gene_name'], gencode['gene_type'], gencode['gene_level'] = \\\n",
    "zip(*gencode.attribute.apply(lambda x: gene_info(x)))\n",
    "\n",
    "gencode.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "gencode = gencode.rename(columns={'seqname': 'chr'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['IG_C_gene', 'IG_C_pseudogene', 'IG_D_gene', 'IG_J_gene',\n",
       "       'IG_J_pseudogene', 'IG_V_gene', 'IG_V_pseudogene', 'IG_pseudogene',\n",
       "       'Mt_rRNA', 'Mt_tRNA', 'TEC', 'TR_C_gene', 'TR_D_gene', 'TR_J_gene',\n",
       "       'TR_J_pseudogene', 'TR_V_gene', 'TR_V_pseudogene', 'artifact',\n",
       "       'lncRNA', 'miRNA', 'misc_RNA', 'processed_pseudogene',\n",
       "       'protein_coding', 'pseudogene', 'rRNA', 'rRNA_pseudogene',\n",
       "       'ribozyme', 'sRNA', 'scRNA', 'scaRNA', 'snRNA', 'snoRNA',\n",
       "       'transcribed_processed_pseudogene',\n",
       "       'transcribed_unitary_pseudogene',\n",
       "       'transcribed_unprocessed_pseudogene',\n",
       "       'translated_processed_pseudogene',\n",
       "       'translated_unprocessed_pseudogene', 'unitary_pseudogene',\n",
       "       'unprocessed_pseudogene', 'vault_RNA'], dtype=object)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(gencode['gene_type'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['CDS', 'exon', 'five_prime_UTR', 'gene', 'start_codon',\n",
       "       'stop_codon', 'stop_codon_redefined_as_selenocysteine',\n",
       "       'three_prime_UTR', 'transcript'], dtype=object)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(gencode['feature'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>parent</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000290825.1;gene_id=ENSG00000290825.1...</td>\n",
       "      <td>N/A</td>\n",
       "      <td>ENSG00000290825.1</td>\n",
       "      <td>DDX11L2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000456328.2;Parent=ENSG00000290825.1;...</td>\n",
       "      <td>ENSG00000290825.1</td>\n",
       "      <td>ENSG00000290825.1</td>\n",
       "      <td>DDX11L2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:1;Parent=ENST0000045...</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>ENSG00000290825.1</td>\n",
       "      <td>DDX11L2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:2;Parent=ENST0000045...</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>ENSG00000290825.1</td>\n",
       "      <td>DDX11L2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:3;Parent=ENST0000045...</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>ENSG00000290825.1</td>\n",
       "      <td>DDX11L2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr  source     feature  start    end score strand frame  \\\n",
       "0  chr1  HAVANA        gene  11869  14409     .      +     .   \n",
       "1  chr1  HAVANA  transcript  11869  14409     .      +     .   \n",
       "2  chr1  HAVANA        exon  11869  12227     .      +     .   \n",
       "3  chr1  HAVANA        exon  12613  12721     .      +     .   \n",
       "4  chr1  HAVANA        exon  13221  14409     .      +     .   \n",
       "\n",
       "                                           attribute             parent  \\\n",
       "0  ID=ENSG00000290825.1;gene_id=ENSG00000290825.1...                N/A   \n",
       "1  ID=ENST00000456328.2;Parent=ENSG00000290825.1;...  ENSG00000290825.1   \n",
       "2  ID=exon:ENST00000456328.2:1;Parent=ENST0000045...  ENST00000456328.2   \n",
       "3  ID=exon:ENST00000456328.2:2;Parent=ENST0000045...  ENST00000456328.2   \n",
       "4  ID=exon:ENST00000456328.2:3;Parent=ENST0000045...  ENST00000456328.2   \n",
       "\n",
       "             gene_id gene_name gene_type  gene_level  \n",
       "0  ENSG00000290825.1   DDX11L2    lncRNA           2  \n",
       "1  ENSG00000290825.1   DDX11L2    lncRNA           2  \n",
       "2  ENSG00000290825.1   DDX11L2    lncRNA           2  \n",
       "3  ENSG00000290825.1   DDX11L2    lncRNA           2  \n",
       "4  ENSG00000290825.1   DDX11L2    lncRNA           2  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gencode.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "gencode['gene_id2'] = gencode['gene_id'].str.split('.').str[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "gencode_gene_only = gencode[gencode['feature'] == 'gene'].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "gencode_gene_only = gencode_gene_only.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>parent</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_level</th>\n",
       "      <th>gene_id2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000290825.1;gene_id=ENSG00000290825.1...</td>\n",
       "      <td>N/A</td>\n",
       "      <td>ENSG00000290825.1</td>\n",
       "      <td>DDX11L2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000290825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>12010</td>\n",
       "      <td>13670</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000223972.6;gene_id=ENSG00000223972.6...</td>\n",
       "      <td>N/A</td>\n",
       "      <td>ENSG00000223972.6</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000227232.5;gene_id=ENSG00000227232.5...</td>\n",
       "      <td>N/A</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000278267.1;gene_id=ENSG00000278267.1...</td>\n",
       "      <td>N/A</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSG00000278267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>29554</td>\n",
       "      <td>31109</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000243485.5;gene_id=ENSG00000243485.5...</td>\n",
       "      <td>N/A</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000243485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62698</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>14149</td>\n",
       "      <td>14673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000198695.2;gene_id=ENSG00000198695.2...</td>\n",
       "      <td>N/A</td>\n",
       "      <td>ENSG00000198695.2</td>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSG00000198695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62699</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>14674</td>\n",
       "      <td>14742</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000210194.1;gene_id=ENSG00000210194.1...</td>\n",
       "      <td>N/A</td>\n",
       "      <td>ENSG00000210194.1</td>\n",
       "      <td>MT-TE</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSG00000210194</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62700</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>14747</td>\n",
       "      <td>15887</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000198727.2;gene_id=ENSG00000198727.2...</td>\n",
       "      <td>N/A</td>\n",
       "      <td>ENSG00000198727.2</td>\n",
       "      <td>MT-CYB</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSG00000198727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62701</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000210195.2;gene_id=ENSG00000210195.2...</td>\n",
       "      <td>N/A</td>\n",
       "      <td>ENSG00000210195.2</td>\n",
       "      <td>MT-TT</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSG00000210195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62702</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000210196.2;gene_id=ENSG00000210196.2...</td>\n",
       "      <td>N/A</td>\n",
       "      <td>ENSG00000210196.2</td>\n",
       "      <td>MT-TP</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSG00000210196</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>62703 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chr   source feature  start    end score strand frame  \\\n",
       "0      chr1   HAVANA    gene  11869  14409     .      +     .   \n",
       "1      chr1   HAVANA    gene  12010  13670     .      +     .   \n",
       "2      chr1   HAVANA    gene  14404  29570     .      -     .   \n",
       "3      chr1  ENSEMBL    gene  17369  17436     .      -     .   \n",
       "4      chr1   HAVANA    gene  29554  31109     .      +     .   \n",
       "...     ...      ...     ...    ...    ...   ...    ...   ...   \n",
       "62698  chrM  ENSEMBL    gene  14149  14673     .      -     .   \n",
       "62699  chrM  ENSEMBL    gene  14674  14742     .      -     .   \n",
       "62700  chrM  ENSEMBL    gene  14747  15887     .      +     .   \n",
       "62701  chrM  ENSEMBL    gene  15888  15953     .      +     .   \n",
       "62702  chrM  ENSEMBL    gene  15956  16023     .      -     .   \n",
       "\n",
       "                                               attribute parent  \\\n",
       "0      ID=ENSG00000290825.1;gene_id=ENSG00000290825.1...    N/A   \n",
       "1      ID=ENSG00000223972.6;gene_id=ENSG00000223972.6...    N/A   \n",
       "2      ID=ENSG00000227232.5;gene_id=ENSG00000227232.5...    N/A   \n",
       "3      ID=ENSG00000278267.1;gene_id=ENSG00000278267.1...    N/A   \n",
       "4      ID=ENSG00000243485.5;gene_id=ENSG00000243485.5...    N/A   \n",
       "...                                                  ...    ...   \n",
       "62698  ID=ENSG00000198695.2;gene_id=ENSG00000198695.2...    N/A   \n",
       "62699  ID=ENSG00000210194.1;gene_id=ENSG00000210194.1...    N/A   \n",
       "62700  ID=ENSG00000198727.2;gene_id=ENSG00000198727.2...    N/A   \n",
       "62701  ID=ENSG00000210195.2;gene_id=ENSG00000210195.2...    N/A   \n",
       "62702  ID=ENSG00000210196.2;gene_id=ENSG00000210196.2...    N/A   \n",
       "\n",
       "                 gene_id    gene_name                           gene_type  \\\n",
       "0      ENSG00000290825.1      DDX11L2                              lncRNA   \n",
       "1      ENSG00000223972.6      DDX11L1  transcribed_unprocessed_pseudogene   \n",
       "2      ENSG00000227232.5       WASH7P              unprocessed_pseudogene   \n",
       "3      ENSG00000278267.1    MIR6859-1                               miRNA   \n",
       "4      ENSG00000243485.5  MIR1302-2HG                              lncRNA   \n",
       "...                  ...          ...                                 ...   \n",
       "62698  ENSG00000198695.2       MT-ND6                      protein_coding   \n",
       "62699  ENSG00000210194.1        MT-TE                             Mt_tRNA   \n",
       "62700  ENSG00000198727.2       MT-CYB                      protein_coding   \n",
       "62701  ENSG00000210195.2        MT-TT                             Mt_tRNA   \n",
       "62702  ENSG00000210196.2        MT-TP                             Mt_tRNA   \n",
       "\n",
       "       gene_level         gene_id2  \n",
       "0               2  ENSG00000290825  \n",
       "1               2  ENSG00000223972  \n",
       "2               2  ENSG00000227232  \n",
       "3               3  ENSG00000278267  \n",
       "4               2  ENSG00000243485  \n",
       "...           ...              ...  \n",
       "62698           3  ENSG00000198695  \n",
       "62699           3  ENSG00000210194  \n",
       "62700           3  ENSG00000198727  \n",
       "62701           3  ENSG00000210195  \n",
       "62702           3  ENSG00000210196  \n",
       "\n",
       "[62703 rows x 15 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gencode_gene_only"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>parent</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_level</th>\n",
       "      <th>gene_id2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>61898</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>152574677</td>\n",
       "      <td>152698700</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000287394.1;gene_id=ENSG00000287394.1...</td>\n",
       "      <td>N/A</td>\n",
       "      <td>ENSG00000287394.1</td>\n",
       "      <td>MAGEA3-DT</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000287394</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        chr  source feature      start        end score strand frame  \\\n",
       "61898  chrX  HAVANA    gene  152574677  152698700     .      -     .   \n",
       "\n",
       "                                               attribute parent  \\\n",
       "61898  ID=ENSG00000287394.1;gene_id=ENSG00000287394.1...    N/A   \n",
       "\n",
       "                 gene_id  gene_name gene_type  gene_level         gene_id2  \n",
       "61898  ENSG00000287394.1  MAGEA3-DT    lncRNA           2  ENSG00000287394  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gencode_gene_only[gencode_gene_only['gene_name']=='MAGEA3-DT']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "unique_gene_id = np.unique(gencode_gene_only['gene_id2'].tolist())\n",
    "\n",
    "repeat_XY_gene_id = []\n",
    "\n",
    "for gene_id in unique_gene_id:\n",
    "    \n",
    "    temp_df = gencode_gene_only[gencode_gene_only['gene_id2'] == gene_id]\n",
    "    \n",
    "    if len(temp_df) > 1:\n",
    "        repeat_XY_gene_id.append(gene_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "47"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(repeat_XY_gene_id)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Update Overlap Funciton"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "def percent_overlap(input_start, input_end, gene_start, gene_end):\n",
    "    \n",
    "    o  = min(input_end, gene_end) - max(input_start, gene_start) + 1    \n",
    "    perc = np.round(o/float(int(gene_end)-int(gene_start)+1)*100, 2)\n",
    "    \n",
    "    return perc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "def overlap_func_sv_to_gene_updated(row, ind, sv_type, sample_id, gencode_df, output_df):\n",
    "    \n",
    "    temp_output_df = output_df.copy()\n",
    "    \n",
    "    temp_df = gencode_df[gencode_df['chr'] == row['CHROM']]\n",
    "    temp_df = temp_df[(temp_df['end'] >= row['POS']) & (temp_df['start'] <= row['END'])]\n",
    "    \n",
    "    \n",
    "    if len(temp_df) != 0:\n",
    "        \n",
    "        for temp_ind, temp_row in temp_df.iterrows():\n",
    "            \n",
    "            if (temp_row['gene_id2'] in repeat_XY_gene_id) and (temp_row['chr'] == 'chrY'):\n",
    "                \n",
    "                pass\n",
    "            \n",
    "            elif temp_row['gene_id2'] in temp_output_df['gene_id2'].tolist():\n",
    "                \n",
    "                temp_output_df.loc[temp_output_df['gene_id2'] == temp_row['gene_id2'], 'count'] += 1\n",
    "                temp_output_df.loc[temp_output_df['gene_id2'] == temp_row['gene_id2'], 'sv_range'].values[0].append((row['POS'], row['END']))\n",
    "                temp_output_df.loc[temp_output_df['gene_id2'] == temp_row['gene_id2'], 'sv_ind'].values[0].append(temp_ind)\n",
    "                temp_output_df.loc[temp_output_df['gene_id2'] == temp_row['gene_id2'], 'percent_overlap'].values[0].append(percent_overlap(row['POS'], row['END'], temp_row['start'], temp_row['end']))\n",
    "                temp_output_df.loc[temp_output_df['gene_id2'] == temp_row['gene_id2'], 'sv_call_id'].values[0].append(row['ID'])\n",
    "                temp_output_df.loc[temp_output_df['gene_id2'] == temp_row['gene_id2'], 'sv_qual'].values[0].append(row['QUAL'])\n",
    "                temp_output_df.loc[temp_output_df['gene_id2'] == temp_row['gene_id2'], 'sv_len'].values[0].append(row['SVLEN'])\n",
    "                temp_output_df.loc[temp_output_df['gene_id2'] == temp_row['gene_id2'], 'sv_GT'].values[0].append(row['GT'])\n",
    "                temp_output_df.loc[temp_output_df['gene_id2'] == temp_row['gene_id2'], 'sv_hap_af'].values[0].append(row['HAP_ALLELIC_FRAC'])\n",
    "                temp_output_df.loc[temp_output_df['gene_id2'] == temp_row['gene_id2'], 'sv_af'].values[0].append(row['ALLELIC_FRAC'])\n",
    "                temp_output_df.loc[temp_output_df['gene_id2'] == temp_row['gene_id2'], 'sv_dist_to_blood'].values[0].append(row['DIST'])\n",
    "\n",
    "            else:\n",
    "                \n",
    "                temp_row['count'] = 1\n",
    "                temp_row['sv_range'] = [(row['POS'], row['END'])]\n",
    "                temp_row['sv_ind'] = [temp_ind]\n",
    "                temp_row['percent_overlap'] = [percent_overlap(row['POS'], row['END'], temp_row['start'], temp_row['end'])]\n",
    "                temp_row['sv_call_id'] = [row['ID']]\n",
    "                temp_row['sv_qual'] = [row['QUAL']]\n",
    "                temp_row['sv_len'] = [row['SVLEN']]\n",
    "                temp_row['sv_GT'] = [row['GT']]\n",
    "                temp_row['sv_hap_af'] = [row['HAP_ALLELIC_FRAC']]\n",
    "                temp_row['sv_af'] = [row['ALLELIC_FRAC']]\n",
    "                temp_row['sv_dist_to_blood'] = [row['DIST']]\n",
    "                temp_row['sv_type'] = sv_type\n",
    "                temp_row['sample_id'] = sample_id\n",
    "                \n",
    "                temp_output_df = temp_output_df.append(temp_row, ignore_index=True)\n",
    "    \n",
    "    \n",
    "    return temp_output_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Small dels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Load filtered somatic small deletions ###\n",
    "\n",
    "##\n",
    "somatic_small_del_path = '/Users/ryanyutian/Desktop/new_panel_TRI_Brain_glioma_sv_processed/somatic_panel_final/small_dels'\n",
    "\n",
    "somatic_small_del_filtered_df_names = []\n",
    "\n",
    "os.chdir(somatic_small_del_path)\n",
    "temp_files = sorted([i for i in os.listdir(somatic_small_del_path) if 'DS' not in i])\n",
    "\n",
    "for file_name in temp_files:\n",
    "    \n",
    "    globals()[file_name[:-4]] = pd.read_csv(file_name)\n",
    "    somatic_small_del_filtered_df_names.append(file_name[:-4])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Currently analyzing: A_RR_GBM809\n",
      "Currently analyzing: A_R_GBM607\n",
      "Currently analyzing: B_P_GBM593\n",
      "Currently analyzing: B_R_GBM898\n",
      "Currently analyzing: C_P_GBM577\n",
      "Currently analyzing: C_R_GBM625\n",
      "Currently analyzing: E_RR_GBM937\n",
      "Currently analyzing: E_R_GBM781\n",
      "Currently analyzing: F_P_GBM620\n",
      "Currently analyzing: F_R_GBM691\n",
      "Currently analyzing: G_P_GBM454\n",
      "Currently analyzing: G_R_GBM833\n",
      "Currently analyzing: H_P_GBM460\n",
      "Currently analyzing: H_R_GBM492\n",
      "Currently analyzing: I_P_GBM440\n",
      "Currently analyzing: I_R_GBM532\n",
      "Currently analyzing: J_P_GBM401\n",
      "Currently analyzing: J_RR_GBM551\n",
      "Currently analyzing: J_R_GBM498\n",
      "Currently analyzing: K_P_GBM529\n",
      "Currently analyzing: K_R_GBM832\n",
      "Currently analyzing: L_P_GBM618\n",
      "Currently analyzing: L_R_SMTB152\n",
      "Currently analyzing: M_P_GBM672\n",
      "Currently analyzing: M_R_GBM828\n",
      "Currently analyzing: N_P_BT2013110\n",
      "Currently analyzing: N_R_GBM745\n",
      "Currently analyzing: O_P_GBM703\n",
      "Currently analyzing: O_R_SMTB781\n",
      "Currently analyzing: P_P_SMTB123\n",
      "Currently analyzing: P_R_SMTB265\n",
      "Currently analyzing: Q_P_SMTB665\n",
      "Currently analyzing: Q_R_GBM1070\n",
      "Currently analyzing: X_R_GBM945\n",
      "Currently analyzing: X_R_SMTB135\n",
      "Currently analyzing: X_R_SMTB241\n",
      "Currently analyzing: X_R_SMTB302\n",
      "Currently analyzing: X_R_SMTB814\n"
     ]
    }
   ],
   "source": [
    "small_del_gene_list_names = []\n",
    "\n",
    "for small_del_df_name in somatic_small_del_filtered_df_names:\n",
    "    \n",
    "    print('Currently analyzing: ' + small_del_df_name[:-41])\n",
    "    temp_df = globals()[small_del_df_name]\n",
    "    temp_output_df = pd.DataFrame(columns=['chr', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', \\\n",
    "                                           'attribute', 'parent', 'gene_id', 'gene_name', 'gene_type', \\\n",
    "                                           'gene_level', 'gene_id2', 'count', 'sv_range', 'sv_ind', \\\n",
    "                                           'percent_overlap', 'sv_call_id', 'sv_qual', 'sv_len', 'sv_GT', \\\n",
    "                                           'sv_hap_af', 'sv_af', 'sv_dist_to_blood', 'sv_type', 'sample_id'])\n",
    "    \n",
    "    for index, row in temp_df.iterrows():\n",
    "        \n",
    "        temp_output_df = overlap_func_sv_to_gene_updated(row, index, 'sdel', small_del_df_name[:-41], gencode_gene_only, temp_output_df)    \n",
    "    \n",
    "    temp_gene_list_name = small_del_df_name[:-41] + '_small_del_gene_list'\n",
    "    small_del_gene_list_names.append(temp_gene_list_name)\n",
    "    \n",
    "    globals()[temp_gene_list_name] = temp_output_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "small_del_gene_list_path = '/Users/ryanyutian/Desktop/new_panel_TRI_Brain_gene_list/all/small_dels/'\n",
    "\n",
    "for df_name in small_del_gene_list_names:\n",
    "    \n",
    "    temp_df = globals()[df_name]\n",
    "    temp_df.to_csv((small_del_gene_list_path + df_name + '.csv'), index=False, sep=',')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Large dels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Load filtered somatic large deletions ###\n",
    "\n",
    "##\n",
    "somatic_large_DEL_path = '/Users/ryanyutian/Desktop/new_panel_TRI_Brain_glioma_sv_processed/somatic_panel_final/large_svs/DEL'\n",
    "\n",
    "somatic_large_DEL_filtered_df_names = []\n",
    "\n",
    "os.chdir(somatic_large_DEL_path)\n",
    "temp_files = sorted([i for i in os.listdir(somatic_large_DEL_path) if 'DS' not in i])\n",
    "\n",
    "for file_name in temp_files:\n",
    "    \n",
    "    globals()[file_name[:-4]] = pd.read_csv(file_name)\n",
    "    somatic_large_DEL_filtered_df_names.append(file_name[:-4])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "large_DEL_gene_list_names = []\n",
    "\n",
    "for large_DEL_df_name in somatic_large_DEL_filtered_df_names:\n",
    "    \n",
    "    print('Currently analyzing: ' + large_DEL_df_name[:-44])\n",
    "    temp_df = globals()[large_DEL_df_name]\n",
    "    temp_output_df = pd.DataFrame(columns=['chr', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', \\\n",
    "                                           'attribute', 'parent', 'gene_id', 'gene_name', 'gene_type', \\\n",
    "                                           'gene_level', 'gene_id2', 'count', 'sv_range', 'sv_ind', \\\n",
    "                                           'percent_overlap', 'sv_call_id', 'sv_qual', 'sv_len', 'sv_GT', \\\n",
    "                                           'sv_hap_af', 'sv_af', 'sv_dist_to_blood', 'sv_type', 'sample_id'])\n",
    "    \n",
    "    for index, row in temp_df.iterrows():\n",
    "        \n",
    "        temp_output_df = overlap_func_sv_to_gene_updated(row, index, 'DEL', large_DEL_df_name[:-44], gencode_gene_only, temp_output_df)    \n",
    "    \n",
    "    temp_gene_list_name = large_DEL_df_name[:-44] + '_large_DEL_gene_list'\n",
    "    large_DEL_gene_list_names.append(temp_gene_list_name)\n",
    "    \n",
    "    globals()[temp_gene_list_name] = temp_output_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "large_DEL_gene_list_path = '/Users/ryanyutian/Desktop/new_panel_TRI_Brain_gene_list/all/large_svs/DEL/'\n",
    "\n",
    "for df_name in large_DEL_gene_list_names:\n",
    "    \n",
    "    temp_df = globals()[df_name]\n",
    "    temp_df.to_csv((large_DEL_gene_list_path + df_name + '.csv'), index=False, sep=',')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Large dups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Load filtered somatic large duplications ###\n",
    "\n",
    "##\n",
    "somatic_large_DUP_path = '/Users/ryanyutian/Desktop/new_panel_TRI_Brain_glioma_sv_processed/somatic_panel_final/large_svs/DUP'\n",
    "\n",
    "somatic_large_DUP_filtered_df_names = []\n",
    "\n",
    "os.chdir(somatic_large_DUP_path)\n",
    "temp_files = sorted([i for i in os.listdir(somatic_large_DUP_path) if 'DS' not in i])\n",
    "\n",
    "for file_name in temp_files:\n",
    "    \n",
    "    globals()[file_name[:-4]] = pd.read_csv(file_name)\n",
    "    somatic_large_DUP_filtered_df_names.append(file_name[:-4])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "large_DUP_gene_list_names = []\n",
    "\n",
    "for large_DUP_df_name in somatic_large_DUP_filtered_df_names:\n",
    "    \n",
    "    print('Currently analyzing: ' + large_DUP_df_name[:-44])\n",
    "    temp_df = globals()[large_DUP_df_name]\n",
    "    temp_output_df = pd.DataFrame(columns=['chr', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', \\\n",
    "                                           'attribute', 'parent', 'gene_id', 'gene_name', 'gene_type', \\\n",
    "                                           'gene_level', 'gene_id2', 'count', 'sv_range', 'sv_ind', \\\n",
    "                                           'percent_overlap', 'sv_call_id', 'sv_qual', 'sv_len', 'sv_GT', \\\n",
    "                                           'sv_hap_af', 'sv_af', 'sv_dist_to_blood', 'sv_type', 'sample_id'])\n",
    "    \n",
    "    for index, row in temp_df.iterrows():\n",
    "        \n",
    "        temp_output_df = overlap_func_sv_to_gene_updated(row, index, 'DUP', large_DUP_df_name[:-44], gencode_gene_only, temp_output_df)    \n",
    "    \n",
    "    temp_gene_list_name = large_DUP_df_name[:-44] + '_large_DUP_gene_list'\n",
    "    large_DUP_gene_list_names.append(temp_gene_list_name)\n",
    "    \n",
    "    globals()[temp_gene_list_name] = temp_output_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "large_DUP_gene_list_path = '/Users/ryanyutian/Desktop/new_panel_TRI_Brain_gene_list/all/large_svs/DUP/'\n",
    "\n",
    "for df_name in large_DUP_gene_list_names:\n",
    "    \n",
    "    temp_df = globals()[df_name]\n",
    "    temp_df.to_csv((large_DUP_gene_list_path + df_name + '.csv'), index=False, sep=',')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Large invs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Load filtered somatic large inversions ###\n",
    "\n",
    "##\n",
    "somatic_large_INV_path = '/Users/ryanyutian/Desktop/new_panel_TRI_Brain_glioma_sv_processed/somatic_panel_final/large_svs/INV'\n",
    "\n",
    "somatic_large_INV_filtered_df_names = []\n",
    "\n",
    "os.chdir(somatic_large_INV_path)\n",
    "temp_files = sorted([i for i in os.listdir(somatic_large_INV_path) if 'DS' not in i])\n",
    "\n",
    "for file_name in temp_files:\n",
    "    \n",
    "    globals()[file_name[:-4]] = pd.read_csv(file_name)\n",
    "    somatic_large_INV_filtered_df_names.append(file_name[:-4])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "large_INV_gene_list_names = []\n",
    "\n",
    "for large_INV_df_name in somatic_large_INV_filtered_df_names:\n",
    "    \n",
    "    print('Currently analyzing: ' + large_INV_df_name[:-44])\n",
    "    temp_df = globals()[large_INV_df_name]\n",
    "    temp_output_df = pd.DataFrame(columns=['chr', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', \\\n",
    "                                           'attribute', 'parent', 'gene_id', 'gene_name', 'gene_type', \\\n",
    "                                           'gene_level', 'gene_id2', 'count', 'sv_range', 'sv_ind', \\\n",
    "                                           'percent_overlap', 'sv_call_id', 'sv_qual', 'sv_len', 'sv_GT', \\\n",
    "                                           'sv_hap_af', 'sv_af', 'sv_dist_to_blood', 'sv_type', 'sample_id'])\n",
    "    \n",
    "    for index, row in temp_df.iterrows():\n",
    "        \n",
    "        temp_output_df = overlap_func_sv_to_gene_updated(row, index, 'INV', large_INV_df_name[:-44], gencode_gene_only, temp_output_df)    \n",
    "    \n",
    "    temp_gene_list_name = large_INV_df_name[:-44] + '_large_INV_gene_list'\n",
    "    large_INV_gene_list_names.append(temp_gene_list_name)\n",
    "    \n",
    "    globals()[temp_gene_list_name] = temp_output_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "large_INV_gene_list_path = '/Users/ryanyutian/Desktop/new_panel_TRI_Brain_gene_list/all/large_svs/INV/'\n",
    "\n",
    "for df_name in large_INV_gene_list_names:\n",
    "    \n",
    "    temp_df = globals()[df_name]\n",
    "    temp_df.to_csv((large_INV_gene_list_path + df_name + '.csv'), index=False, sep=',')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
